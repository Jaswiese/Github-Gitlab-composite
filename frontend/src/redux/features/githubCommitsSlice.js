/* eslint-disable max-len */
import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
/**
 * getGithubCommits asyncThunk created
 * used to retrieve the github commits from the backend api
 * two params are passed to the async function
 * both params are used within the endpoint construction
 * data is parsed from json and returned
 */
export const getGithubCommits = createAsyncThunk( 'githubCommits/getGithubCommits',
    async ({username, repo}) => {
      const response = await fetch(`/github/${username}/${repo}`);
      const data = response.json();
      return data;
    },
);
/**
 * githubCommitSlice declared
 * state named: githubCommit
 * initial state declared with,
 * githubCommits as an empty array
 * status as an empty string
 * error as null
 */
export const githubCommitsSlice = createSlice({
  name: 'githubCommits',
  initialState: {
    githubCommits: [],
    status: '',
    error: null,
  },
  /**
   * extraReducers declared handling promise state of the asyncThunk
   * .pending
   * status set to loading
   * .fulfilled
   * status set to success
   * githubCommits set to the payload from the thunk
   * .rejected
   * status set to rejected
   * error set to payload
   */
  extraReducers: {
    [getGithubCommits.pending]: (state) => {
      state.status = 'loading';
    },
    [getGithubCommits.fulfilled]: (state, action) => {
      state.status = 'success';
      state.githubCommits = action.payload;
    },
    [getGithubCommits.rejected]: (state, action) => {
      state.status = 'rejected';
      state.error = action.payload;
    },
  },
});
// slice reducer is exported
export default githubCommitsSlice.reducer;
