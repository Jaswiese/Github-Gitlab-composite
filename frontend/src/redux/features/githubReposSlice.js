import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
/**
 * getGithubRepos asyncthunk created
 * async function has username as param
 * response assigned the result of the fetch to the backend endpoint
 * data assigned the value of the response parsed from json and returned
 */
export const getGithubRepos = createAsyncThunk('githubRepos/getGithubRepos',
    async (username) => {
      const response = await fetch(`/github/repos/${username}`);
      const data = response.json();
      return data;
    });
/**
 * githubRepoSlice declared
 * with default state,
 * githubRepos - empty array
 * status - empty string
 * error - null
 * ExtraReducers, handling asyncThunk promises
 * .pending
 * status is set to loading
 * .fulfilled
 * status is set to success
 * githubRepos is set to payload (data from thunk)
 * .rejected
 * status is set to rejected
 * error is set to payload (data from thunk)
 */
const githubReposSlice = createSlice({
  name: 'githubRepos',
  initialState: {
    githubRepos: [],
    status: '',
    error: null,
  },
  extraReducers: {
    [getGithubRepos.pending]: (state) => {
      state.status = 'loading';
    },
    [getGithubRepos.fulfilled]: (state, action) => {
      state.status = 'success';
      state.githubRepos = action.payload;
    },
    [getGithubRepos.rejected]: (state, action) => {
      state.status = 'rejected';
      state.error = action.payload;
    },
  },
});
// slice's reducer is exported
export default githubReposSlice.reducer;
