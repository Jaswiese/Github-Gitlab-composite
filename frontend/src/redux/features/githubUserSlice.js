import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
/**
 * getGithubUser asyncThunk declared
 * Retrieves the github user
 * async function has one param username
 * response assigned the value of the fetch call to the backend api
 * data assigned the parsed response from json and returned
 */
export const getGithubUser = createAsyncThunk('githubUser/getGithubUser',
    async (username) => {
      const response = await fetch(`/github/${username}`);
      const data = response.json();
      return data;
    });
/**
 * githubUserSlice created
 * default initial state declared,
 * githubUser - empty array
 * status - empty string
 * error as null value
 * ExtraReducers (thunks promises handled)
 * .pending
 * sets status to loading
 * .fulfilled
 * sets status to success
 * githubUser is assigned the value of the payload
 * .rejected
 * sets status to rejected
 * error set to payload
 */
const githubUserSlice = createSlice({
  name: 'githubUser',
  initialState: {
    githubUser: [],
    status: '',
    error: null,
  },
  extraReducers: {
    [getGithubUser.pending]: (state) => {
      state.status = 'loading';
    },
    [getGithubUser.fulfilled]: (state, action) => {
      state.status = 'success';
      state.githubUser = action.payload;
    },
    [getGithubUser.rejected]: (state, action) => {
      state.status = 'rejected';
      state.error = action.payload;
    },
  },
});
// slices reducer exported
export default githubUserSlice.reducer;
