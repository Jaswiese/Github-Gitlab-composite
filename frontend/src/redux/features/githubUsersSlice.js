import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
/**
 * getGithubUsers asyncThunk created
 * retrieves the list of github users
 * async has one param username used within fetch call
 * data assigned the parsed response from json value
 */
export const getGithubUsers = createAsyncThunk('githubUsers/getGithubUsers',
    async (username) => {
      const response = await fetch(`/github/search/${username}`);
      const data = response.json();
      return data;
    });
/**
 * githubUsersSlice created
 * default state declared,
 * githubUsers - empty array
 * status -  empty string
 * error - null value
 * ExtraReducers (asyncThunk promise handler)
 * .pending
 * status set to loading
 * .fulfilled
 * status set to success
 * githubUsers assigned the value of the payload (data)
 * .rejected
 * status set to rejected
 * error set to the payload (data)
 */
const githubUsersSlice = createSlice({
  name: 'githubUsers',
  initialState: {
    githubUsers: [],
    status: '',
    error: null,
  },
  extraReducers: {
    [getGithubUsers.pending]: (state) =>{
      state.status = 'loading';
    },
    [getGithubUsers.fulfilled]: (state, action) => {
      state.status = 'success';
      state.githubUsers = action.payload;
    },
    [getGithubUsers.rejected]: (state, action) => {
      state.status = 'rejected';
      state.error = action.payload;
    },
  },
});
// reducer exported
export default githubUsersSlice.reducer;
