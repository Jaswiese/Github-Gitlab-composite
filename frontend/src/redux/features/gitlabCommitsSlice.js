/* eslint-disable max-len */
import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
/**
 * gitlabCommits asyncThunk declared
 * used to return gitlab projects commits
 * Async func has two params username and projectid used with the fetch request
 * response is assigned the fetch requests return value
 * data is assigned the parsed request from json and returned
 */
export const getGitlabCommits = createAsyncThunk('gitlabCommit/getGitlabCommits',
    async ({username, projectid}) => {
      const response = await fetch(`/gitlab/${username}/${projectid}`);
      const data = response.json();
      return data;
    },
);
/**
 * gitlabCommitSlice declared
 * default initial state declared
 * gitlabCommits - empty array
 * status - empty string
 * error - null value
 * ExtraReducers - (handles thunk)
 * .pending
 * status set to loading
 * .fulfilled
 * .status set to success
 * gitlabCommits set to the payload
 * .rejected status set to rejected
 * error set to the payload
 */
const gitlabCommitSlice = createSlice({
  name: 'gitlabCommits',
  initialState: {
    gitlabCommits: [],
    status: '',
    error: null,
  },
  extraReducers: {
    [getGitlabCommits.pending]: (state) => {
      state.status = 'loading';
    },
    [getGitlabCommits.fulfilled]: (state, action) => {
      state.status = 'success';
      state.gitlabCommits = action.payload;
    },
    [getGitlabCommits.rejected]: (state, action) => {
      state.status = 'rejected';
      state.error = action.payload;
    },
  },
});
// reducer exported
export default gitlabCommitSlice.reducer;
