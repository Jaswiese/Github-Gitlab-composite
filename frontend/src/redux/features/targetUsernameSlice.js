import {createSlice} from '@reduxjs/toolkit';
// default initial state declared
const initialState = {
  username: '',
};
// targetUsername slice declared (handles user search input state)
export const targetUsernameSlice = createSlice({
  name: 'targetUsername',
  initialState,
  reducers: {
    // setTargetUsername action - sets username to payload value
    setTargetUsername: (state, action) => {
      state.username = action.payload;
    },
  },
});
// action exported
export const {setTargetUsername} = targetUsernameSlice.actions;
// slices reducer exported
export default targetUsernameSlice.reducer;
