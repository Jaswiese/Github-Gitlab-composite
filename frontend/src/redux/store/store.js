// configureStore imported from toolkit - used to consolidate different state
import {configureStore} from '@reduxjs/toolkit';
// all reducers are imported from corresponding slices
import githubReducer from '../features/githubUsersSlice';
import targetUsernameReducer from '../features/targetUsernameSlice';
import gitlabUsersReducer from '../features/gitlabUsersSlice';
import githubUserReducer from '../features/githubUserSlice';
import githubCommitsReducer from '../features/githubCommitsSlice';
import githubReposReducer from '../features/githubReposSlice';
import gitlabProjReducer from '../features/gitlabProjectsSlice';
import gitlabCommitsReducer from '../features/gitlabCommitsSlice';

/** ConfigureStore - configure Redux store
 * all reducers consolidated and exported as a single store
 */
export default configureStore({
  reducer: {
    githubUsers: githubReducer,
    targetUsername: targetUsernameReducer,
    gitlabUsers: gitlabUsersReducer,
    githubUser: githubUserReducer,
    githubCommits: githubCommitsReducer,
    githubRepos: githubReposReducer,
    gitlabProj: gitlabProjReducer,
    gitlabCommits: gitlabCommitsReducer,
  },
});
