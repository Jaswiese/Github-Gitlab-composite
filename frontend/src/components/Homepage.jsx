/* eslint-disable max-len */
import React from 'react';
// bootstrap components imported
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Spinner from 'react-bootstrap/Spinner';
// redux hooks imported
import { useDispatch, useSelector } from 'react-redux';
// users component imported
import Users from './Users';
// slice actions imported
import { setTargetUsername } from '../redux/features/targetUsernameSlice';
import { getGithubUsers } from '../redux/features/githubUsersSlice';
import { getGitlabUsers } from '../redux/features/gitlabUsersSlice';
// css imported
import '../resources/css/homepage.css';
/**
 * Homepage component
 * styles from homepage.css and bootstrap components used for layout
 * bootstrap also used for layout styling
 * Displays the search bar and github repos and gitlab projects as children
 * @return {*} displays the homepage to the user
 */
function Homepage() {
  // useDispatch declared as dispatch variable
  const dispatch = useDispatch();
  // state status synced used for display logic
  const gitHubUsersStatus = useSelector((state) => state.githubUsers.status);
  const gitlabUsersStatus = useSelector((state) => state.gitlabUsers.status);
  // targetUsername retrieved with the useSelector hook
  const targetUsername = useSelector((state) => state.targetUsername.username);
  // lab & hub users state retrieved with useSelector hook
  const githubUsers = useSelector((state) => state.githubUsers.githubUsers);
  const gitlabUsers = useSelector((state) => state.gitlabUsers.gitlabUsers);
  // need to add search functions and handling of data here
  /**
   * onTargetUserNameChange function
   * handles user input into the search bar
   * calls dispatch to fire the setTargetUsername action,
   * with the event value as payload
   * @param {*} e input event
   */
  const onTargetUserNameChange = (e) => {
    dispatch(setTargetUsername(e.target.value));
  };
  /**
   * search function
   * calls dispatch to fire the getGithubUsers & getGitlabUsers actions,
   * with username param as payload
   * @param {*} username param used for redux actions
   */
  const search = (username) => {
    dispatch(getGithubUsers(username));
    dispatch(getGitlabUsers(username));
  };
  /**
   * displayGithub function
   * used as a display helper function, depending on the status of the fetch state
   * If the data is still loading or undefined the function returns a Spinner
   * If the status is success (data is fetched)
   * an array of users is created
   * the function then returns Users components with a map function
   * assigning the props on each iteration
   * by default or in the case of an error,
   * an appropriate heading is returned
   * @param {*} status gitHubUsersStatus
   * @return {*} html for github profiles
   */
  const displayGithub = (status) => {
    if (status === 'loading' || status === 'undefined') {
      return (
        <Spinner animation="border" role="status">
          <span className="visually-hidden">Loading...</span>
        </Spinner>
      );
    }
    if (status === 'success') {
      const arrOfGithubUsers = githubUsers.items;

      return (
        <>
          <h4 className="text-center">Github user profiles</h4>
          <Container fluid className="users-container-home">
            {arrOfGithubUsers.map((user) => (
              <Users
                key={user.id}
                id={user.id}
                img={user.avatar_url}
                name={user.login}
                platform="github"
              />
            ))}
          </Container>
        </>
      );
    }
    return (
      <h4 className="text-center pt-5">
        Search for github users profiles to be displayed
      </h4>
    );
  };
  /**
   * displayGitlab function
   * helper function to aid in display depending on application state
   * if the gitlab data is still being retrieved,
   * a spinner is displayed
   * if the data is successfully retrieved,
   * A map function is called to return Users component with props,
   * on each iteration.
   * Default state
   * a message displayed to prompt the user to search
   * @param {*} status gitlab users data state
   * @return {*} html to display gitlab users
   */
  const displayGitlab = (status) => {
    if (status === 'loading' || status === 'undefined') {
      return (
        <Spinner animation="border" role="status">
          <span className="visually-hidden">Loading...</span>
        </Spinner>
      );
    }
    if (status === 'success') {
      return (
        <>
          <h4 className="text-center pt-5">Gitlab user profile(s)</h4>
          <Container fluid className="users-container-home">
            {gitlabUsers.map((user) => (
              <Users
                key={user.id}
                id={user.id}
                name={user.username}
                img={user.avatar_url}
                platform="gitlab"
              />
            ))}
          </Container>
        </>
      );
    }
    return (
      <h4 className="text-center pt-5">
        Search for gitlab users profiles to be displayed
      </h4>
    );
  };
  /**
   * Primary return of the Homepage component
   * BootStrap components & css used for layout
   * on the search bar (text input) field,
   * value is set to the targetUsername variable
   * onChange the onTargetUserNameChange function is called
   * on the search button,
   * a test id is added
   * onclick event is added that fires the search function
   * at the bottom both display functions are called,
   * with their corresponding status variables as params
   */
  return (
    <>
      <Container fluid>
        <Row className="justify-content-center">
          <Col className="col-md-6">
            <h1 className="text-center">Git Composite Searchüîç</h1>
          </Col>
        </Row>
        <Row className="justify-content-center pt-2">
          <Col className="col-md-8 justify-content-center">
            <div className="search-bar-container d-flex justify-content-center">
              <label htmlFor="searchBar" className="search-label" />
              <input
                type="text"
                className="search-bar"
                placeholder="search github/gitlab username here"
                name="search-bar"
                id="search-bar"
                value={targetUsername}
                onChange={onTargetUserNameChange}
              />
              <button
                data-testid="search-button"
                type="button"
                className="btn btn-primary"
                onClick={() => {
                  search(targetUsername);
                }}
              >
                Search
              </button>
            </div>
          </Col>
        </Row>
      </Container>
      {displayGithub(gitHubUsersStatus)}
      {displayGitlab(gitlabUsersStatus)}
    </>
  );
}

export default Homepage;
