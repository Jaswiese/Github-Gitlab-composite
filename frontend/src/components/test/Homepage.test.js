/* eslint-disable no-undef */
import React from 'react';
// function from the testing library imported
import {render, screen, fireEvent} from '@testing-library/react';
// dependency components and functions imported
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import store from '../../redux/store/store';
import Homepage from '../Homepage';
import {setTargetUsername} from '../../redux/features/targetUsernameSlice';
import {getGithubUsers} from '../../redux/features/githubUsersSlice';

// custom component render declared and redux store passed (due to the use of useParams hook Router wrapped).
const renderHomepage = () =>
  render(
      <Provider store={store}>
        <Router>
          <Homepage />
        </Router>
      </Provider>,
  );
  /**
   * unit test added to see if the getGithubUsers action is working correctly
   * check if the correct user login is reflected in the redux state
   */
test('getGithubUsers', async ()=>{
  store.dispatch(setTargetUsername('svedova'));
  const test = 'svedova';
  let state = null;
  await store.dispatch(getGithubUsers(test)).then(()=>{
    state = store.getState();
  });
  const target = state.githubUsers.githubUsers.items[0].login;
  expect(target).toBe('svedova');
});
/**
 * Testing the Homepage component change with a snapshot test
 * simulates a user clicked the search button after the user has populated the search bar
 */
test('testing if the homepage displays user details after search', async () => {
  store.dispatch(setTargetUsername('svedova'));
  const {rerender, asFragment} = renderHomepage();
  rerender(
      <Provider store={store}>
        <Router>
          <Homepage />
        </Router>
      </Provider>,
  );
  const searchButton = screen.getByTestId('search-button');
  fireEvent.click(searchButton);
  const results = await screen.findByText('Username: svedova');
  expect(results).toHaveTextContent('svedova');
  expect(asFragment()).toMatchSnapshot();
});

