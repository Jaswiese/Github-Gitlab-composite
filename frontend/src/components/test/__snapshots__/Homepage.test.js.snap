// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing if the homepage displays user details after search 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <div
      class="justify-content-center row"
    >
      <div
        class="col-md-6 col"
      >
        <h1
          class="text-center"
        >
          Git Composite Searchüîç
        </h1>
      </div>
    </div>
    <div
      class="justify-content-center pt-2 row"
    >
      <div
        class="col-md-8 justify-content-center col"
      >
        <div
          class="search-bar-container d-flex justify-content-center"
        >
          <label
            class="search-label"
            for="searchBar"
          />
          <input
            class="search-bar"
            id="search-bar"
            name="search-bar"
            placeholder="search github/gitlab username here"
            type="text"
            value="svedova"
          />
          <button
            class="btn btn-primary"
            data-testid="search-button"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </div>
  </div>
  <h4
    class="text-center"
  >
    Github user profiles
  </h4>
  <div
    class="users-container-home container-fluid"
  >
    <div
      class="outer-user-container-home"
    >
      <img
        alt="profile-picture"
        class="profile-picture-home"
        src="https://avatars.githubusercontent.com/u/3321893?v=4"
      />
      <h4
        class="user-name-home"
      >
        Username: svedova
      </h4>
      <button
        class="btn btn-primary view-button-home"
        type="button"
      >
        view
      </button>
    </div>
    <div
      class="outer-user-container-home"
    >
      <img
        alt="profile-picture"
        class="profile-picture-home"
        src="https://avatars.githubusercontent.com/u/88555532?v=4"
      />
      <h4
        class="user-name-home"
      >
        Username: svedovatto
      </h4>
      <button
        class="btn btn-primary view-button-home"
        type="button"
      >
        view
      </button>
    </div>
    <div
      class="outer-user-container-home"
    >
      <img
        alt="profile-picture"
        class="profile-picture-home"
        src="https://avatars.githubusercontent.com/u/20651407?v=4"
      />
      <h4
        class="user-name-home"
      >
        Username: Svedovatem48
      </h4>
      <button
        class="btn btn-primary view-button-home"
        type="button"
      >
        view
      </button>
    </div>
    <div
      class="outer-user-container-home"
    >
      <img
        alt="profile-picture"
        class="profile-picture-home"
        src="https://avatars.githubusercontent.com/u/54397660?v=4"
      />
      <h4
        class="user-name-home"
      >
        Username: svedova-tutti
      </h4>
      <button
        class="btn btn-primary view-button-home"
        type="button"
      >
        view
      </button>
    </div>
    <div
      class="outer-user-container-home"
    >
      <img
        alt="profile-picture"
        class="profile-picture-home"
        src="https://avatars.githubusercontent.com/u/57945246?v=4"
      />
      <h4
        class="user-name-home"
      >
        Username: svedova-test
      </h4>
      <button
        class="btn btn-primary view-button-home"
        type="button"
      >
        view
      </button>
    </div>
  </div>
  <div
    class="spinner-border"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;
